File: src/move.cpp
#include <iostream>
#include <sstream>
#include <csignal>

#include "ros/ros.h"
#include "std_msgs/String.h"
#include "bittleet/MoveServo.h"

#include <pigpiod_if2.h>

class Move{
public:
    Move(int pi, int handle) : 
            _srv(_n.advertiseService("bittleet_move/move", &Move::command_callback, this)),
            _pi(pi),
            _h(handle){}

    void run() {
        ros::spin();
    }

    bool command_callback(bittleet::MoveServo::Request &req, bittleet::MoveServo::Response &res) {
        if (req.index != 0 && (req.index < 8 || req.index > 15)) {
            ROS_WARN("Index %d invalid", (int)req.index);
            return false;
        }
        if (req.angle < -90.0 || req.angle > 90.0) {
            ROS_WARN("Angle %f out of range", req.angle);
            return false;
        }

        char command[30];
        int n = 0;
        n = sprintf(command, "m %d %.0f ", req.index, req.angle);
        ROS_INFO("Sending command '%s'", command);

        if (n > 0) {
            _write_to_serial(command, n);
        } else {
            return false;
        }
        time_sleep(0.1);
        char response[128] = {0};
        int len = _serial_read(response, 128);

        res.result=std::string(response);

        return true;

    }

private:
    void _write_to_serial(char* bytes, int len){
        int result  = serial_write(_pi, _h, bytes, (unsigned)len);
        if (result != 0) {
            ROS_ERROR("Serial write failed: %d ", result);
        }
    }

    int _serial_read(char *bytes, int max_len) {
        int index = 0;
        int byte = serial_read_byte(_pi, _h);

        while ((byte >= 0) && (index < max_len)) {
            bytes[index] = (char)byte;
            index++;       
            byte = serial_read_byte(_pi, _h);
        }
        return index;
    }

    ros::NodeHandle _n;
    ros::ServiceServer _srv;

    int _pi = 0;
    int _h = 0;
};

static int pi = -1;
static int handle = -1;

void shutdown() {
    std::cout << "Shutting down connection to pigpiod\n";
     if (handle >= 0) {
        serial_close(pi, handle);
    }
    if (pi >= 0) {
        pigpio_stop(pi);
    }
}

void at_signal(int i) {
   shutdown();
   exit(0);
}

int main(int argc, char *argv[]) {
    std::signal(SIGINT, at_signal);

    pi = pigpio_start("localhost", "8888");
    if (pi < 0) {
        std::cout << "Connecting to pigpiod failed\n";
        return 1;
    }

    char serialPort[11] = "/dev/ttyS0";
    handle = serial_open(pi, serialPort, 115200, 0);
    if (handle < 0) {
        std::cout << "Serial port unavaliable\n";
        shutdown();
        return 1;
    }

    ros::init(argc, argv, "bittleet_move");

    Move node(pi, handle);
    node.run();

    shutdown();
    return 0;
}

File: scripts/move.py
#!/usr/bin/env python

import pigpio
import time

import rospy
from bittleet.srv import MoveServo, MoveServoRequest, MoveServoResponse

class Move:
    JOINT_INDEXES = [0, 8, 9, 10, 11, 12, 13, 14, 15]

    def __init__(self):
        self.pi = pigpio.pi()
        if not self.pi.connected:
            return
        self.handle = self.pi.serial_open("/dev/ttyS0", 115200)

        self.node = rospy.init_node('bittleet_move')

    def __del__(self):
        if self.handle != None:
            self.pi.serial_close(self.handle)

    def run(self):
        if not self.pi.connected:
            return
        self.srv = rospy.Service(rospy.get_name() +'/move', MoveServo, self.handle_move)
        rospy.spin()

    def handle_move(self, req: MoveServoRequest):
        angle = req.angle
        if not -90.0 <= angle <= 90.0:
            raise rospy.ServiceException("angle {} out of range".format(angle))
        index = req.index
        if index not in Move.JOINT_INDEXES:
            raise rospy.ServiceException("index {} invalid".format(index))

        done = False
        res = MoveServoResponse()
        command = f"m {index} {angle:.0f} "
        print(f"Sending '{command}'")
        
        while not done:
            self.pi.serial_write(self.handle, command)

            time.sleep(0.05)
            len, data = self.pi.serial_read(self.handle, 100)

            if len > 0:
                try:
                    res.result = data.decode("utf-8")
                    done = True
                except:
                    pass
            else:
                res.result = "<No Reply>"
                done = True


        return res


if __name__ == "__main__":
    try:
        move = Move()
        move.run()
    except rospy.ROSInterruptException:
        pass

